docs: Module 1 and 2 (multiple Sections) (B9lab content updates 1) (#791)
cosmos/sdk-tutorials: b9lab-content/3-main-concepts/04-accounts.md

---
title: "Accounts"
order: 4
description: Accounts, addresses and keys in Cosmos.
---

# Accounts

## What is an Account?

> An account is a pair of keys called PubKey, a public key, and PrivKey, a private key. A public is a unique identifier for a user or entity that is safe to disclose.  

Reviewer: ```suggestion
An account is a pair of keys called PubKey (a public key), and PrivKey (a private key). A public is a unique identifier for a user or entity that is safe to disclose.  
```

Suggestion for clarity.

Private keys are sensitive information that users are required to manage confidentially. Private keys are used to sign information in a way that **proves** to others that the message was signed by someone using the private key that corresponds to a given public key. This is done without revealing the private key itself.

## Public Key Cryptography

> Modern cryptographic systems leverage computer capabilities to make accessible the power of certain mathematical functions. Public key cryptography, also known as **asymmetril cryptography**, is a cryptographic system that employs pairs of keys. Every pair consists of a public and a private key. The public key can be shared publicly and the security of the system is not endangered as long as the private key is not disclosed.

Reviewer 2: ```suggestion
Modern cryptographic systems leverage computer capabilities to make accessible the power of certain mathematical functions. Public key cryptography, also known as **asymmetric cryptography**, is a cryptographic system that employs pairs of keys. Every pair consists of a public and a private key. The public key can be shared publicly and the security of the system is not endangered as long as the private key is not disclosed.
```

Reviewer 2: We should probably be running a spell checker in the CI

* **Encryption**, because only the private key can decrypt the information encrypted with the public key.

We will focus on the Authentication aspect.

Compared to symmetric key algorithms, asymmetric ones do not require parties to use a secure channel to exchange the keys for encryption and decryption.

Public key cryptography assures confidentiality, authenticity and non-repudiability. Examples of applications include [S/MIME](https://en.wikipedia.org/wiki/S/MIME) and [GPG](https://en.wikipedia.org/wiki/GNU_Privacy_Guard), as well as the basis of several Internet standards like [SSL](https://www.ssl.com/faqs/faq-what-is-ssl/) and [TLS](https://en.wikipedia.org/wiki/Transport_Layer_Security).

Due to its computational complexity, asymmetric cryptography is normally applied to small data blocks. In a hybrid system, symmetric and asymmetric cryptography are combined. For example, asymmetric encryption could be employed to transfer a symmetric encryption, which would then be used as an encryption key for the message. Examples of this hybrid system include [PGP](https://en.wikipedia.org/wiki/Pretty_Good_Privacy).

> The length of keys is vital. Asymmetric cryptographic keys are usually very long. One can keep in mind a general principle: the longer the key, the more difficult it is to break the code. To break an asymmetric key with a brute force attack, the attacker needs to try every possible key. The longer the key is, the more difficult it is to guess.

Reviewer: ```suggestion
The length of keys is vital. Asymmetric cryptographic keys are usually very long. One can keep in mind a general principle: the longer the key, the more difficult it is to break the code. To break an asymmetric key with a brute force attack, the attacker would need to try every possible key.
```
Redundant sentence.

### Sign and Verify Example

Alice wants to make sure that Bob's public announcement is indeed from Bob:

* Bob gives Alice his public key
* Bob signs his announcement with his private key
* Bob sends Alice his announcement and its signature
* Alice verifies the signature with Bob's public key

When Alice has verified the signature, she will see that the announcement was signed using the private key that corresponds to Bob’s public key which is already known to represent Bob.

In summary, private keys are used to **prove** that messages originate from owners of accounts that are known by their public keys. More precisely, signatures **prove** that messages were signed by someone with knowledge of the private key that corresponds to a given public key. This is the basis of user authentication in a blockchain. For this reason, private keys are jealously guarded secrets.

> ## Hyper-Deterministic Wallets

Reviewer 2: ```suggestion
## Hierarchical-Deterministic Wallets
```

Blockchains generally maintain ledgers of user accounts and rely on public key cryptography for user authentication. Implicitly, knowledge of one’s public and private keys is a requirement for executing transactions. Client software applications known as wallets provide methods to generate new key pairs and store them between sessions, as well as basic services such as creating transactions, signing messages, interacting with applications (for example, web pages) and communicating with the blockchain.

Although it is technically feasible to generate and store multiple key pairs in a wallet, key management is less than ideal for users in that scenario. Given that all keys would exist in one place only, users would need to devise ways to recover those keys in adverse situations such as the loss or destruction of the computer. To put it simply, the more accounts, the more keys to back up.

> A Hyper-deterministic Wallet uses a single seed phrase to generate many key pairs. In this way, only the seed phrase needs to be backed up.

Reviewer 2: ```suggestion
A Hierarchical-Deterministic (HD) Wallet uses a single seed phrase to generate many key pairs. In this way, only the seed phrase needs to be backed up.
```

Authentication is implemented as signature verification.

* Users generate transactions, sign transactions and send the signed transaction to the blockchain.
* Transactions are formatted with the public key as part of the message. Signatures are verified by confirming that the signature’s public key matches the public key in the message. If the message is signed by anyone other than the purported sender, then the signature is invalid and the transaction is rejected.

In case the foregoing is unclear, consider the following pseudo message:

```
Message: {
      Payload: {
  Sender: “0x1234”,
  Data: “Hello World”}
      Signature: “0xabcd”
}
```
Passing `Payload` into the signature verification function will return a sender. The derived sender must match the sender in the `Payload` itself. This will confirm that the `Payload` could only originate from someone with knowledge of the private key corresponding to `Sender: “0x1234”`

## Signature Schemes

As you might guess, there is more than one implementation of the public key signature process described so far. Cosmos SDK supports the following digital key schemes for creating digital signatures:

> * [secp256k1](https://www.secg.org/sec2-v2.pdf), as implemented in the SDK's crypto/keys/secp256k1 package.

Reviewer 2: Should probably mention that most use secp256k1 and this is the same as Bitcoin ?
